sgl-kernel/csrc/moe/moe_fused_gate.cu

```diff
@@ -210,21 +214,38 @@ __device__ void moe_fused_gate_impl(
         indices_ptr[idx] = static_cast<int32_t>(expert);
       }
 
-      // accumulate sum
-      if (thread_group_idx == 0) {
+      // accumulate sum for first k-1 elements only
+      if (thread_group_idx == 0 && k_idx < topk - 1) {
         output_sum += output_ptr[idx];
       }
     }
 
     __syncthreads();
   }
 
+  // Handle n_share_experts_fusion if needed
+  if (n_share_experts_fusion > 0 && thread_group_idx == 0) {
+    int64_t last_idx = topk * thread_row + (topk - 1);
+
+    // Use round-robin to select expert
+    int64_t expert_offset = thread_row % n_share_experts_fusion;
+    indices_ptr[last_idx] = static_cast<int32_t>(params.NUM_EXPERTS + expert_offset);
+
+    // Set the weight to the sum of the first k-1 weights divided by routed_scaling_factor
+    output_ptr[last_idx] = output_sum / routed_scaling_factor;
+
+    __syncthreads();
+  } else if (thread_group_idx == 0) {
+    // If not using shared experts, add the last weight to output_sum
+    output_sum += output_ptr[topk * thread_row + (topk - 1)];
+  }
```

Comment:
About `////////////////////// Topk //////////////////////` section: it seems we are still spending time computing real_topk+num_fusion things. For example, for llama4, it may be top2 instead of top1.

Thus what about

* Suppose we remove name `topk`, and have `topk_excluding_share_expert_fusion` or `topk_with_shared_expert_fusion` etc (to avoid users getting confused which one is which one)
* In the `////////////////////// Topk //////////////////////` section above, we use the `topk_excluding_share_expert_fusion` to compute everything (thus only top1 in llama4)
* Then here we do not need this else if as well